src/game.cpp

```diff
@@ -578,54 +629,193 @@ void test_game() //!OCLINT tests may be many
+
+    const int first_player_index = get_collision_members(g)[1];
+    const int second_player_index = get_collision_members(g)[0];
+    const int loser_index = get_losing_player_index(g, first_player_index, second_player_index);
+
+    assert(loser_index == 1);
+
+    // After a tick, passive player does not shrink
+    const int inv_player_size_before =  get_nth_player_size(g, loser_index);
+    g.tick();
+    const int inv_player_size_after =  get_nth_player_size(g, loser_index);
     assert(inv_player_size_after == inv_player_size_before);
   }
   #endif
 
-  //#define FIX_ISSUE_463
+
+  #define FIX_ISSUE_463
   #ifdef FIX_ISSUE_463
-  // Players lose invulnerability after a short time
+  // Can get and set a player's passive state
+  {
+    game g;
+    player& p = g.get_player(0);
+
+    // Check default duration
+    assert(p.get_duration_passive() == 2000);
+
+    int new_duration = 200;
+    p.set_duration_passive(new_duration);
+    assert(p.get_duration_passive() == new_duration);
+  }
+
+  // Game cannot increment a player's passive timer unless it is passive
+  {
+    game g;
+    player& p = g.get_player(0);
+    assert(!is_passive(p));
+    assert(p.get_passive_timer() == 0);
+
+    g.increment_passive_timers();
+    assert(p.get_passive_timer() == 0);
+
+    int timer_before = p.get_passive_timer();
+    p.set_state_passive();
+    g.increment_passive_timers();
+    int timer_after = p.get_passive_timer();
+    int difference = timer_after - timer_before;
+    assert(difference == 1);
+  }
+
+  // A player's passive timer cannot be bigger than its passive duration,
+  // Otherwise it will be reset to 0
+  {
+    game g;
+    player& p = g.get_player(0);
+    g.set_player_state_passive(p);
+    int duration = p.get_duration_passive();
+    for (int i = 0; i < duration - 1; i++)
+    {
+      assert(p.get_state() == player_state::passive);
+      int timer_before = p.get_passive_timer();
+      g.tick();
+      int timer_after = p.get_passive_timer();
+      int difference = timer_after - timer_before;
+      assert(difference == 1);
+      assert(timer_after <= duration);
+    }
+    g.tick();
+    assert(!p.get_passive_timer());
+    int timer = p.get_passive_timer();
+    assert(timer == 0);
+  }
+
+  // Game cannot reset a player's passive timer if it's less than duration
+  {
+    game g;
+    player& p = g.get_player(0);
+
+    int timer = p.get_passive_timer();
+    assert(timer == 0);
+
+    p.set_state_passive();
+    int duration = p.get_duration_passive();
+
+    for (int i = 0; i < duration - 1; i++)
+    {
+      int timer_before = p.get_passive_timer();
+      g.tick();
+      int timer_after = p.get_passive_timer();
+      int difference = timer_after - timer_before;
+      assert(difference == 1);
+      assert(timer_after <= duration);
+      timer_before = p.get_passive_timer();
+      g.reset_passive_timers();
+      timer_after = p.get_passive_timer();
+      assert(timer_after == timer_before);
+    }
+  }
```

Comment:
Same as above, I don't think it's necessary to explicitly test for the timer?
Unless you think the issue needed to be broken down in smaller steps - in which case, sure, go ahead and create smaller intermediate issues towards losing invulnerability after some time :)

Reply:
I added these tests for debugging purposes as I encountered quite a lot of issues. I was making sure that all the parts I had added were working as I expected.

Reply:
The main test (if it's good!) should already ensure that everything works as expected. If players lose invulnerability after some time and not earlier, as #463 demands, then we have our feature and we're happy!
I'd argue we only care about the underlying mechanisms and potential issues *if* a problem indeed manifests after completing the test. Then it's time to make a new issue to signal the problem, and write a test that ensures it is fixed.

That being said, using tests as debug towards completing another test is entirely fine, but then no need to leave it in the code once you have resolved the final test!

Reply:
> The main test (if it's good!) should already ensure that everything works as expected. If players lose invulnerability after some time and not earlier, as #463 demands, then we have our feature and we're happy! I'd argue we only care about the underlying mechanisms and potential issues _if_ a problem indeed manifests after completing the test. Then it's time to make a new issue to signal the problem, and write a test that ensures it is fixed.
> 
> That being said, using tests as debug towards completing another test is entirely fine, but then no need to leave it in the code once you have resolved the final test!

Yes that's true, I mean I would add extra debugging tests besides the main tests and remove them right before merging, this could be a better way than what was done in this PR 😉